### PR TITLE
Separate AccountLocks + Refactor

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
- Have been doing incremental improvements to `AccountLocks` structure and functionality

#### Summary of Changes
- Move `AccountLocks` to its' own module
- Refactor/rename top-level functions
- Add testing of **only** account locks structure

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts.rs [hunk 1]
```diff
@@ -561,48 +509,6 @@ impl Accounts {
...
-                account_locks.write_locks.insert(*k);
-            } else if !account_locks.lock_readonly(k) {
-                account_locks.insert_new_readonly(k);
-            }
-        }
-
-        Ok(())
-    }
-
-    fn unlock_account<'a>(
```
Comment (apfitzge) [inline accounts-db/src/accounts.rs:592]: these functions used to be on `Accounts` but don't even use `self`.
They are no on the `AccountLocks` struct directly.

**Comment by apfitzge** [comment]
Separating this out and making fields non-pub makes a follow-up change to use a single map simpler.
That change will reduce the number of pubkey hashes we need to do.

File: accounts-db/src/account_locks.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        !self.is_locked_write(key)
+    }
+
+    fn can_write_lock(&self, key: &Pubkey) -> bool {
+        // If the key is not read-locked or write-locked, it can be write-locked
+        !self.is_locked_readonly(key) && !self.is_locked_write(key)
+    }
+
+    fn lock_readonly(&mut self, key: &Pubkey) {
+        *self.readonly_locks.entry(*key).or_default() += 1;
```
Comment (tao-stones) [inline accounts-db/src/account_locks.rs:82]: this is much better than `if !lock_readonly() then insert_new_readonly()` 🎉

File: accounts-db/src/account_locks.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    /// Returns an error if any of the accounts are already locked in a way
+    /// that conflicts with the requested lock.
+    pub fn try_lock_accounts_for_transaction<'a>(
+        &mut self,
+        keys: impl Iterator<Item = (&'a Pubkey, bool)> + Clone,
+    ) -> Result<(), TransactionError> {
+        for (key, writable) in keys.clone() {
+            if writable {
+                if !self.can_write_lock(key) {
+                    debug!("Writable account in use: {:?}", key);
```
Comment (tao-stones) [inline accounts-db/src/account_locks.rs]: nit: maybe remove this debug line, doubt it is monitored at all.

File: accounts-db/src/account_locks.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    write_locks: AHashSet<Pubkey>,
+    readonly_locks: AHashMap<Pubkey, u64>,
+}
+
+impl AccountLocks {
+    /// Lock the account keys in `keys` for a transaction.
+    /// The bool in the tuple indicates if the account is writable.
+    /// Returns an error if any of the accounts are already locked in a way
+    /// that conflicts with the requested lock.
+    pub fn try_lock_accounts_for_transaction<'a>(
```
Comment (tao-stones) [inline accounts-db/src/account_locks.rs]: nitpick on naming: the function say do something "_for_transaction", but there is no transaction in parameter list, only account keys. So maybe just "try_lock_accounts()"?

File: accounts-db/src/account_locks.rs (hunk 3 above)
Comment (apfitzge) [reply accounts-db/src/account_locks.rs]: 16d492d9e0b260403d853c9e5fe972c82648ef71

File: accounts-db/src/account_locks.rs (hunk 4 above)
Comment (apfitzge) [reply accounts-db/src/account_locks.rs]: a8182f81db8d84f7e10ea017650d36d0c733e227

**Review by tao-stones — APPROVED** [review]
lgtm

**Pull request merged**